File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if(head == null) return null;
        // create a copy list without thinking about random pointer
        // interweave both the lists so that random is node.next ... you know
        Node tmp = head;
        while(tmp!=null){
            Node newNode = new Node(tmp.val);
            newNode.next = tmp.next;
            tmp.next = newNode;
            tmp = newNode.next;
        }
        
        // now iterate and assign the random pointers
        tmp = head;
        while(tmp!=null){
            if(tmp.random != null){
                tmp.next.random = tmp.random.next; 
            }else{
                tmp.next.random = null;
            }
            tmp = tmp.next.next;
        }

        // separate the lists
        tmp = head;
        Node copy = tmp.next;
        Node result = copy;
        while(true){
            tmp.next = copy.next;
            if(copy.next == null) break;
            copy.next = copy.next.next;
            tmp = tmp.next;
            copy = copy.next;
        }

        return result;
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        """
        [[3,null],[3,0],[3,null]]
        [3,null]->[3,null]->[3,null]
        0 : [3,null]
        1 : [3,null]
        2 : [3,null]
        """
        if head is None:
            return None
        
        # Create a new list
        prev = Node(head.val)
        newHead = prev
        curr = head.next
        while curr:
            prev.next = Node(curr.val)
            prev = prev.next
            curr = curr.next

        # Put new list in map based on index as keys
        # Put original list also in map but index as values
        index = {}
        tmp = newHead
        curr = head
        idx = 0
        while tmp:
            index[idx] = tmp
            index[curr] = idx
            idx += 1
            tmp = tmp.next
            curr = curr.next

        curr = head
        tmp = newHead
        while curr:
            if curr.random is None:
                pass
            else:
                tmp.random = index[index[curr.random]]
            curr = curr.next
            tmp = tmp.next

        return newHead

        
